python/ray/data/datasource/filename_provider.py

```diff
@@ -0,0 +1,80 @@
+from typing import Any, Dict, Optional
+
+from ray.data.block import Block
+from ray.util.annotations import PublicAPI
+
+
+@PublicAPI(stability="alpha")
+class FilenameProvider:
+    """Generates filenames when you write a :class:`~ray.data.Dataset`.
+
+    Use this class to customize what your written files look like.
+
+    Some methods write each row to a separate file, while others write each block to a
+    separate file. For example, :meth:`ray.data.Dataset.write_images` writes individual
+    rows, and :func:`ray.data.Dataset.write_parquet` writes blocks of data.
+
+    If you're writing each row to a separate file, implement
+    :meth:`~FilenameProvider.get_filename_for_row`. Otherwise, implement
+    :meth:`~FilenameProvider.get_filename_for_block`.
+
+    Example:
+
+        .. testcode::
+
+            from ray.data.datasource import FilenameProvider
+
+            class LabeledFilenameProvider(FilenameProvider):
+
+                def __init__(self, file_format: str):
+                    self.file_format = file_format
+
+                def get_filename_for_row(self, row, file_index):
+                    return f"{row['label']}-{file_index}.{file_format}"
+    """
+
+    def get_filename_for_block(self, block: Block, file_index: int) -> str:
+        """Generate a filename for a block of data.
+
+        .. note::
+            Filenames must be unique and idempotent for a given file index. This means
+            your filenames can't be random, and if you call this method twice with the
+            same arguments, you should get the same filename both times.
+
+        Args:
+            block: The block that will be written to a file.
+            file_index: The index of the file that will be written.
```

Comment:
It would be good to explain this arg a bit more. For example, what value gets passed in here, how it will change if we're writing one row/file vs block/file, etc.

Also, is it guaranteed to be contiguous? If not, file_index is a bit misleading because it sounds like it will go from [0, total_files). I think either we have to provide contiguous file indices or we provide the whole tuple of (task_idx, block_idx) / (task_idx, block_idx, row_idx). Otherwise it may not be usable for cases where you want to preserve some order in the written dataset.